eth/tracers/api.go

```diff
@@ -53,13 +53,6 @@ const (
 	// and reexecute to produce missing historical state necessary to run a specific
 	// trace.
 	defaultTraceReexec = uint64(128)
-
-	// defaultTracechainMemLimit is the size of the triedb, at which traceChain
-	// switches over and tries to use a disk-backed database instead of building
-	// on top of memory.
-	// For non-archive nodes, this limit _will_ be overblown, as disk-backed tries
-	// will only be found every ~15K blocks or so.
-	defaultTracechainMemLimit = common.StorageSize(500 * 1024 * 1024)
```

Comment:
I see that this is now removed. Some context for this thing -- if we do chain tracing from an early block, and want to e.g. trace 1M blocks between 12M and 13M on an archive node. 
Then the way it works is that it will dig up state for 12M, and trace with that. After that, it will continue operating on the same state (now with modifications from that one block). It will continue, using in-memory trie database as state-backend, until it hits the `defaultTracechainMemLimit`. At this point, e.g. `120001000`, it will release that state, and load a new one from disk. 

This means that it is both pretty fast (reuses the in-memory state instead of always loading from disk) and also does not go out of memory (since it doesn't accumulate indefinitely). I implemented this in https://github.com/ethereum/go-ethereum/pull/23736 . 

So, I'm curious, how is this scenario handled in this PR? "Always reload", "never reload" or something else?

Reply:
Now we will always try to find the state in disk first. The logic is here https://github.com/rjl493456442/go-ethereum/blob/improve-chain-tracer/eth/state_accessor.go#L81

I totally agree that we should reuse state in disk as much as possible. So I make it as the default behavior.

Reply:
> I totally agree that we should reuse state in disk as much as possible. So I make it as the default behavior.

...I'm not 100% certain that we're in agreement.. (?)
So what I'm saying is the opposite: we want to avoid starting 'fresh' from the disk-state most of the time. Only do it when memory constraints force us to. 


--- 
Now, I might be ok with doing a cleaner-but-slower approach if it makes the code and the semantics about "state loading" simpler, but if we go down that route, we should definitely measure it first, to see how much slower it becomes.

Reply:
Get it.

For archive node, I think in-memory solution should have a lot of benefits. `cleaner-but-slower` approach doesn't make too much sense. Thanks for pointing it out, I will think about it.

Reply:
I have added it back.

But the behavior is slightly different. Previously whenever the cached nodes reach the threshold, we switch over to `preferDisk` mode. It can be inefficient in archive node which always build the fresh trie database afterwards. Now this mode will only be enabled once the memory usage exceeds the limit.

Please take another look.